[UR] Enable hwloc in UMF linux build

PatKamin opened the pull request:
~~Update UR to statically link UMF with hwloc on Linux.~~
This will enable hwloc support in UMF on Linux.

~~Also, always use the v2.9.3 hwloc version for hwloc builds.~~

PatKamin commented:
> Drafted until the UMF fix is pulled down to llvm via UMF update in UR.

Fix is now pulled down, ready for review.

igchor commented:
You should probably also update https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md (remove hwloc as a dependency that needs to be installed by the user)

PatKamin commented:
> You should probably also update https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md (remove hwloc as a dependency that needs to be installed by the user)

Right, now it will automatically be downloaded - deleted user requirement.

bader commented on sycl/doc/GetStartedGuide.md:
```diff
@@ -44,8 +44,6 @@ and a wide range of compute accelerators such as GPU and FPGA.
 * `python` - [Download](https://www.python.org/downloads/)
 * `ninja` -
 [Download](https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages)
-* `hwloc` version 2.3 or later (Linux only)
-  * libhwloc-dev or hwloc-devel package on linux
```
> Update UR to statically link hwloc v2.9.3 both on linux and Windows.

If I get the change right, the dependency on `hwloc` still exists. The change removed the runtime dependency, but not the compile time dependency. Am I right?

This document describes the environment for DPC++ compiler developers, not DPC++ compiler users.

Your change seems to raise the version requirement. Please, update the `hwloc` version to 2.9.3.

PatKamin replied:
The `hwloc` library will always be fetched from sources. Regardless of `hwloc-dev` packages user has installed, the fetched version will be linked with the UMF library. That's why I remove `hwloc` from this dependencies list, as the user-installed `hwloc-dev` package will not be used by UMF.

bader commented:
> > You should probably also update https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md (remove hwloc as a dependency that needs to be installed by the user)
> 
> Right, now it will automatically be downloaded - deleted user requirement.

If `hwloc` dependencies will be automatically resolved by CMake, please, remove installation of these dependencies from CI scripts. Specifically, from https://github.com/intel/llvm/blob/7cc9e8054f3827c83d4724601b8e7929199cf87f/devops/scripts/install_build_tools.sh#L26 and https://github.com/intel/llvm/blob/7cc9e8054f3827c83d4724601b8e7929199cf87f/.github/workflows/sycl-docs.yml#L36.

PatKamin commented:
> > > You should probably also update https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md (remove hwloc as a dependency that needs to be installed by the user)
> > 
> > 
> > Right, now it will automatically be downloaded - deleted user requirement.
> 
> If `hwloc` dependencies will be automatically resolved by CMake, please, remove installation of these dependencies from CI scripts. Specifically, from
> 
> https://github.com/intel/llvm/blob/7cc9e8054f3827c83d4724601b8e7929199cf87f/devops/scripts/install_build_tools.sh#L26
> 
> and
> https://github.com/intel/llvm/blob/7cc9e8054f3827c83d4724601b8e7929199cf87f/.github/workflows/sycl-docs.yml#L36
> 
> .

Thanks, done

PatKamin commented:
> Do we still need `brew install hwloc` on MacOS build CI workflow: https://github.com/intel/llvm/blob/sycl/.github/workflows/sycl-macos-build-and-test.yml#L33 ?

I've removed it now, it shouldn't be required anymore.

> I recall installing `hwloc` on Windows CI runners. Now that we are fetching `hwloc` from sources and statically linking it, am I good to remove `hwloc` installations on Windows CI runners once this PR gets merged?

`openmp` still needs hwloc, so this can be done with a caution on Windows CI runners that are not used for openmp tests, I think.
> 
> Also, can you please update the PR title? - at the first glace it looked like the change is for Linux only. But the description says that the change is for both Windows and Linux.

I've updated the misleading commit message. `hwloc` already is linked to UMF statically on Windows even without this change.

aarongreig reviewed (approved):
UR LGTM

uditagarwal97 reviewed (approved):
CI changes LGTM.

PatKamin commented:
@bader, I've added an `automake` dependency for building `hwloc`.

bader commented:
> Also, always use the v2.9.3 hwloc version for hwloc builds.

> I've added an automake dependency for building hwloc.

If we always use fixed version of the library, I prefer CMake to use the pre-built binaries. Building it from sources is unnecessary overhead.

WRT adding dependency on `automake` for building `hwloc`, I'd like @AlexeySachkov to approve this.

PatKamin commented:
> If we always use fixed version of the library, I prefer CMake to use the pre-built binaries. Building it from sources is unnecessary overhead.

I can make it a fallback: first search for `hwloc` binaries installed in the system, if not found - fetch `hwloc` from sources. I would also remove specifying the exact `hwloc` version in `llvm` leaving it to `UMF` to decide on version. @bader, what do you think?

bader commented:
> > If we always use fixed version of the library, I prefer CMake to use the pre-built binaries. Building it from sources is unnecessary overhead.
> 
> I can make it a fallback: first search for `hwloc` binaries installed in the system, if not found - fetch `hwloc` from sources. I would also remove specifying the exact `hwloc` version in `llvm` leaving it to `UMF` to decide on version. @bader, what do you think?

I suggest we fetch binaries instead of sources.
https://www.open-mpi.org/software/hwloc/v2.9/ has links to the binaries for different platforms.

`hwloc` handling must be done in UMF CMake files.

PatKamin commented:
> WRT adding dependency on automake for building hwloc, I'd like @AlexeySachkov to approve this.

@AlexeySachkov, is it ok to add `automake` dependency?

PatKamin commented:
> hwloc handling must be done in UMF CMake files.

@bader, I've removed `hwloc` version handling from `llvm`. It's handled by UMF CMake files now.

@bader, @AlexeySachkov, ping on the `automake` dependency.

AlexeySachkov commented:
> > hwloc handling must be done in UMF CMake files.
> 
> @bader, I've removed `hwloc` version handling from `llvm`. It's handled by UMF CMake files now.
> 
> @bader, @AlexeySachkov, ping on the `automake` dependency.

In general, I'm not a fan of having those dependencies because they complicate our SW stack for not so apparent reason. Pardon my ignorance, but why `hwloc` and even UMF have to be used by default?
At least for the amount of testing we perform at intel/llvm - are there any tests which can detect side effects of UMF and/or `hwloc` being present or absent? If not, then I would be strongly against bringing that dependency by default, because it means that its an overhead without a reason.

Also, why do we need to build either from source at all? I understand reason for building UR from sources, because it is moves together with SYCL RT on a very fast pace and changes have to be done in sync, but is it really the case for UMF and `hwloc`?

As @bader suggested, why can't we just download a binary release of `hwloc` (and ideally UMF) during UMF (UR) build?

> I suggest we fetch binaries instead of sources.
> https://www.open-mpi.org/software/hwloc/v2.9/ has links to the binaries for different platforms.

Also tagging @jsji here, because any dependencies we add here will also affects our downstream.

PatKamin commented:
> In general, I'm not a fan of having those dependencies because they complicate our SW stack for not so apparent reason. Pardon my ignorance, but why `hwloc` and even UMF have to be used by default? At least for the amount of testing we perform at intel/llvm - are there any tests which can detect side effects of UMF and/or `hwloc` being present or absent? If not, then I would be strongly against bringing that dependency by default, because it means that its an overhead without a reason.

The UMF library is required for allocation handling by default in UR, the user cannot opt-out, UR can't be built without UMF. There are two ways that UR gets UMF now. UMF is either fetched by UR or the user can force CMake to search for the installed UMF with `find_package` using the `UR_USE_EXTERNAL_UMF` env var.
Although currently `hwloc` can be disabled and not used, the env var that lets the user to do this (`UMF_DISABLE_HWLOC`) is planned to be removed prior the stable `1.0` release of UMF. `hwloc` will be a required dependency of UMF.

> 
> Also, why do we need to build either from source at all? I understand reason for building UR from sources, because it is moves together with SYCL RT on a very fast pace and changes have to be done in sync, but is it really the case for UMF and `hwloc`?

We don't release UMF prebuilt libraries as a part of open-source releases and there are no plans to do so. Users using UMF have to fetch it.

> As @bader suggested, why can't we just download a binary release of `hwloc` (and ideally UMF) during UMF (UR) build?
> 
> > I suggest we fetch binaries instead of sources.
> > https://www.open-mpi.org/software/hwloc/v2.9/ has links to the binaries for different platforms.
> 

I don't fetch `hwloc` binaries as we apply a [patch](https://github.com/oneapi-src/unified-memory-framework/blob/f0a34f69aece73e881c2f003feaa9022fd80a4bf/cmake/fix_coverity_issues.patch#L4) to its sources.
I also wanted the intel/llvm hwloc handling to be the same as on the downstream where I wouldn't like to force having hwloc binaries installed in the system as a requirement.

On the second hand, currently on Linux hwloc is a requirement of intel/llvm and CI seems to be ready for enabling hwloc in UMF Linux builds, I could verify that. So I propose to stick with having hwloc as a requirement in intel/llvm and not fetch it.

vinser52 commented:
AS @PatKamin said today UMF project does not distribute binaries. We can prioritize this discussion and estimate what should be done to distribute UMF in binary form. I agree that it might be a best option.

Regarding why UMF is not optional, @PatKamin said that UR uses UMF to support memory pooling. Another reason is the interop between runtimes. UMF provides observability for other runtimes. For example, MPI uses UMF to get the memory properties of the memory allocated by SYCL applications.

AlexeySachkov commented:
> I don't fetch `hwloc` binaries as we apply a [patch](https://github.com/oneapi-src/unified-memory-framework/blob/f0a34f69aece73e881c2f003feaa9022fd80a4bf/cmake/fix_coverity_issues.patch#L4) to its sources.

Does it mean that we need to redistribute our own version of `hwloc` then?

PatKamin commented:
> Does it mean that we need to redistribute our own version of `hwloc` then?

No, it's just one of the options. The other is to require the user to have hwloc installed.

PatKamin commented:
Reverted back changes with fetching hwloc. Leaving hwloc as a required dependency. Failing CI jobs also fail on other PRs, looks like these failures are not caused by changes in this PR.

AlexeySachkov commented:
Thanks for educating me on the topic, @vinser52, @PatKamin.

> > > I don't fetch `hwloc` binaries as we apply a [patch](https://github.com/oneapi-src/unified-memory-framework/blob/f0a34f69aece73e881c2f003feaa9022fd80a4bf/cmake/fix_coverity_issues.patch#L4) to its sources.
> > 
> > Does it mean that we need to redistribute our own version of `hwloc` then?
>
> No, it's just one of the options. The other is to require the user to have hwloc installed.

What is the point of patching `hwloc` if we link with it dynamically and don't redistribute the updated version? Shouldn't we link statically with `hwloc` then?

Do I understand correctly that we don't fetch `hwloc` on Windows either and it is also expected to be present on a build system?

PatKamin commented:
> What is the point of patching `hwloc` if we link with it dynamically and don't redistribute the updated version? Shouldn't we link statically with `hwloc` then?

The patch is needed to satisfy Coverity requirements in case hwloc is fetched and built as a static library.

> Do I understand correctly that we don't fetch `hwloc` on Windows either and it is also expected to be present on a build system?

In Windows we have to do the fetch and link statically with hwloc always: https://github.com/intel/llvm/blob/sycl/sycl/cmake/modules/FetchUnifiedRuntime.cmake#L129-L130

Please, note that I've minimized the changes in this PR recently. On current sycl branch, hwloc already is a required llvm dependency and is installed in intel/llvm CI environment. I've removed the fetching for Linux builds in favor of using the installed hwloc library to reduce the unnecessary overhead of hwloc fetch and build during llvm build.

This will, however, require changes in the downstream CI environment, so I put this PR on draft until it's ready there.

PatKamin commented:
Changes wait for https://github.com/intel/llvm/pull/17551 to be merged.

PatKamin commented:
@intel/unified-runtime-reviewers, please take a look - all tests passed. Changes are now minimal.

PatKamin commented:
@intel/llvm-gatekeepers, I think this PR is ready to merge

ldrumm commented:
It doesn't look like Alexey's concerns have been addressed so I'm going to hold off merge until he's confirmed he's okay with this.

As a side note: the PR description doesn't do what it says
